### ISSUE TITLE
tainted values become untainted in tied hashes

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3833]
Migrated from [rt.perl.org#6758](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=6758) (status was 'resolved')

Searchable as RT6758$

**Comment by p5pRT** [comment]
### From phanson@well.com



Using a tied hash untaints the values stored in the hash under perl 5\.6\.0 
\(both Solaris 5\.6 and Windows \(Active Perl\)\.  Here's an example program
that demonstrates the problem&#8203;:

&nbsp;    \#\!/usr/local/bin/perl5 \-T

&nbsp;    package IHash ;

&nbsp;    sub check\_tainted
&nbsp;    \{
&nbsp;        my $type  = shift ;
&nbsp;        my $value = shift ;
&nbsp;        my $istainted = \! eval \{ my $x = $value\, kill 0 ; 1 \} ;
&nbsp;        print $istainted ? "TAINTED&#8203;: " : 'not tainted&#8203;: ' ;
&nbsp;        print "$type = $value\\n" ;
&nbsp;    \}

&nbsp;    sub STORE
&nbsp;    \{
&nbsp;        my $this  = shift ;
&nbsp;        my $key   = shift ;
&nbsp;        my $value = shift ;
&nbsp;        check\_tainted\('key'\, $key\) ;
&nbsp;        check\_tainted\('value'\, $value\) ;
&nbsp;    \}

&nbsp;    sub TIEHASH
&nbsp;    \{
&nbsp;        my $class =  shift ;
&nbsp;        my $this  =  \{\} ;
&nbsp;        return bless $this => $class ;
&nbsp;    \}

&nbsp;    package main ;

&nbsp;    my %hash ;

&nbsp;    tie %hash\, 'IHash' ;
&nbsp;    my $key   = shift @&#8203;ARGV ;
&nbsp;    my $value = shift @&#8203;ARGV ;
&nbsp;    IHash&#8203;::check\_tainted\('original key'\, $key\) ;
&nbsp;    IHash&#8203;::check\_tainted\('original value'\, $value\) ;

&nbsp;    $hash\{$key\} = $value ;

When run with arguments of 'arg1' and 'arg2' on the command line\, it
produces the following output&#8203;:

&nbsp;    TAINTED&#8203;: original key = arg1
&nbsp;    TAINTED&#8203;: original value = arg2
&nbsp;    TAINTED&#8203;: key = arg1
&nbsp;    not tainted&#8203;: value = arg2

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl v5.6.0:

Configured by phanson at Wed Sep  6 18:10:35 PDT 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=solaris, osvers=2.6, archname=sun4-solaris
    uname='sunos well 5.6 generic_105181-21 sun4u sparc sunw,ultra-4 '
    config_args=''
    hint=previous, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='gcc', optimize='-O', gccversion=2.8.1
    cppflags='-I/usr/local/BerkeleyDB/include -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    ccflags ='-I/usr/local/BerkeleyDB/include -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    stdchar='unsigned char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags ='-L/usr/local/BerkeleyDB/lib -L/usr/local/lib '
    libpth=/usr/local/BerkeleyDB/lib /usr/local/lib /lib /usr/lib /usr/ccs/lib
    libs=-lsocket -lnsl -ldb -ldl -lm -lc -lcrypt -lsec
    libc=/lib/libc.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl'
    cccdlflags='-fPIC', lddlflags='-W,l -G -L/usr/local/BerkeleyDB/lib -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /usr/local/lib/perl5/5.6.0/sun4-solaris
    /usr/local/lib/perl5/5.6.0
    /usr/local/lib/perl5/site_perl/5.6.0/sun4-solaris
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/p/h/phanson
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=.:/home/p/h/phanson/bin:/usr/local/gnu/bin:/usr/local/lib/acct:/usr/local/bin:/usr/ccs/bin:/usr/bin:/usr/sbin:/usr/ucb:.:/opt/SUNWspro/bin:/usr/well/bin:/usr/local/sbin:/usr/local/rdb/bin:/usr/local/sysadmin/bin:/usr/local/lib/nuoracle
    PERL_BADLANG (unset)
    SHELL=/usr/bin/ksh

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

still present in bleedperl@&#8203;19435

**Comment by p5pRT** [comment]
### From @muir

Still present in 5\.8\.3

**Comment by p5pRT** [comment]
### From @iabyn

This has been fixed by commit b112cff9879ef9e20ee30b1a9ec813b1336a3093
in the branch davem/post\-5\.12 which should be merged back into blead
once 5\.12 has been released\, and thus appear in 5\.14\.

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
